File: native/android/sfiora/src/main/java/com/sandrox/sfiora/NdefWritePolicy.java
package com.sandrox.sfiora;

import java.util.Arrays;

final class NdefWritePolicy {
    private NdefWritePolicy() {
    }

    static void validateInitializationMessage(
            NdefMessage message,
            NdefExternalType marker
    ) {
        int matches = 0;
        for (NdefRecord record : message.getRecords()) {
            if (marker.matches(record)) {
                matches++;
            }
        }
        if (matches != 1) {
            throw new IllegalArgumentException(
                    "The initialization message must contain exactly one "
                            + "record matching the external type marker"
            );
        }
    }

    static void validateWritable(
            boolean writable,
            int capacityBytes,
            NdefMessage message
    ) throws NfcOperationException {
        if (!writable) {
            throw new NfcOperationException(
                    NfcErrorCode.TAG_READ_ONLY,
                    "The detected NDEF tag is read-only",
                    false
            );
        }
        if (message.getByteCount() > capacityBytes) {
            throw new NfcOperationException(
                    NfcErrorCode.NDEF_CAPACITY_EXCEEDED,
                    "The NDEF message requires "
                            + message.getByteCount()
                            + " bytes but the tag capacity is "
                            + capacityBytes
                            + " bytes",
                    false
            );
        }
    }

    static void verify(NdefMessage expected, NdefMessage actual)
            throws NfcOperationException {
        if (actual == null
                || !Arrays.equals(
                expected.getSerializedData(),
                actual.getSerializedData()
        )) {
            throw new NfcOperationException(
                    NfcErrorCode.WRITE_VERIFICATION_FAILED,
                    "The NDEF message read from the tag differs from the message written",
                    true
            );
        }
    }
}

File: native/ios/Sources/Sfiora/NdefWritePolicy.swift
import Foundation

enum NdefWritePolicy {
    static func validateInitializationMessage(
        _ message: NdefMessage,
        marker: NdefExternalType
    ) throws {
        guard message.records.filter(marker.matches).count == 1 else {
            throw NdefWritePolicyError.invalidInitializationMessage
        }
    }

    static func validateWritable(
        status: NfcNdefStatus,
        capacityBytes: Int,
        message: NdefMessage
    ) throws {
        guard status == .readWrite else {
            throw NfcError(
                code: .tagReadOnly,
                message: "The detected NDEF tag is read-only",
                recoverable: false
            )
        }
        guard message.byteCount <= capacityBytes else {
            throw NfcError(
                code: .ndefCapacityExceeded,
                message:
                    "The NDEF message requires \(message.byteCount) bytes "
                    + "but the tag capacity is \(capacityBytes) bytes",
                recoverable: false
            )
        }
    }

    static func verify(expected: NdefMessage, actual: NdefMessage?) throws {
        guard actual?.serializedData == expected.serializedData else {
            throw NfcError(
                code: .writeVerificationFailed,
                message: "The NDEF message read from the tag differs from the message written",
                recoverable: true
            )
        }
    }
}

enum NdefWritePolicyError: Error, Equatable, LocalizedError {
    case invalidInitializationMessage

    var errorDescription: String? {
        "The initialization message must contain exactly one record matching "
            + "the external type marker"
    }
}
